### ISSUE TITLE
Clspv crashes accessing the first array element with -physical-storage-buffers

### ISSUE THREAD

**Issue by BukeBeyond** [google/clspv#1179]
```
struct E { int i1, i2, i3, i4; };

struct S
{
    int Shift; // Fails
    struct E e[3];
};

kernel void Kernel(global struct S* s)
{
    s->e[0] = s->e[1]; // Fails writing e[0]
    // struct E e = s->e[1]; s->e[0] = e; // OK, workaround
    // s->e[1] = s->e[0]; // Fails reading e[0]
    // struct E e = s->e[0]; s->e[1] = e; // OK, workaround
    // s->e[1] = s->e[2]; // OK
    // s->e[2] = s->e[1]; // OK
}
```
https://godbolt.org/z/q75WoKs4T

This could be related to https://github.com/google/clspv/issues/1176 https://github.com/google/clspv/issues/1177 https://github.com/google/clspv/issues/1178 .  However, this is for -physical-storage-buffers, and the int Shift; triggers the bug, unlike others, where the Shift bypasses it.

### Additional observations:

When the compilation is external with Clang, in CLC++, and with strong optimizations, the OK workarounds are not effective.
Clang compiles the first Fail line to the following, which crashes Clspv: 
```
define dso_local spir_kernel void @Kernel(ptr addrspace(1) nocapture noundef align 4 %0) local_unnamed_addr #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_base_type !7 !kernel_arg_type_qual !8 {
  %2 = getelementptr inbounds %struct.S, ptr addrspace(1) %0, i64 0, i32 1
  %3 = getelementptr inbounds %struct.S, ptr addrspace(1) %0, i64 0, i32 1, i64 1
  %4 = addrspacecast ptr addrspace(1) %3 to ptr addrspace(4)
  %5 = addrspacecast ptr addrspace(1) %2 to ptr addrspace(4)
  tail call void @llvm.memcpy.p4.p4.i64(ptr addrspace(4) noundef align 4 dereferenceable(16) %5, ptr addrspace(4) noundef align 4 dereferenceable(16) %4, i64 16, i1 false), !tbaa.struct !9
  ret void
}
```

The second line, OK, workaround, to the following, which also crashes Clspv.
```
define dso_local spir_kernel void @Kernel(ptr addrspace(1) nocapture noundef align 4 %0) local_unnamed_addr #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_base_type !7 !kernel_arg_type_qual !8 {
  %2 = getelementptr inbounds %struct.S, ptr addrspace(1) %0, i64 0, i32 1
  %3 = getelementptr inbounds %struct.S, ptr addrspace(1) %0, i64 0, i32 1, i64 1
  %4 = addrspacecast ptr addrspace(1) %3 to ptr addrspace(4)
  %5 = addrspacecast ptr addrspace(1) %2 to ptr addrspace(4)
  tail call void @llvm.memcpy.p4.p4.i64(ptr addrspace(4) noundef align 4 dereferenceable(16) %5, ptr addrspace(4) noundef align 4 dereferenceable(16) %4, i64 16, i1 false)
  ret void
}
```

And when the struct E size is reduced to a single int i1, memcpy is bypassed, Clspv finally succeeds:
```
define dso_local spir_kernel void @Kernel(ptr addrspace(1) nocapture noundef align 4 %0) local_unnamed_addr #0 !kernel_arg_addr_space !4 !kernel_arg_access_qual !5 !kernel_arg_type !6 !kernel_arg_base_type !7 !kernel_arg_type_qual !8 {
  %2 = getelementptr inbounds %struct.S, ptr addrspace(1) %0, i64 0, i32 1
  %3 = getelementptr inbounds %struct.S, ptr addrspace(1) %0, i64 0, i32 1, i64 1
  %4 = addrspacecast ptr addrspace(1) %3 to ptr addrspace(4)
  %5 = addrspacecast ptr addrspace(1) %2 to ptr addrspace(4)
  %6 = load i32, ptr addrspace(4) %4, align 4, !tbaa !9
  store i32 %6, ptr addrspace(4) %5, align 4, !tbaa !9
  ret void
}
```

In conclusion, Clang compilation in CLC++ extensively emits the generic addrspace(4), even when no function calls or generics are used!  Thus, it is likely necessary for Clspv to default these generics to global addrspace(1) in CLC++ mode.  Clspv does this already with load and store instructions, but seems to have problems when memcpy is involved.
